packages/navie/src/navie.ts

```diff
@@ -46,6 +46,7 @@ export interface ClientRequest {
   question: string;
   codeSelection?: string;
```

Comment:
Historically, this `codeSelection` was the code that the user had selected in their IDE when they opened Navie. So, it was a string. And even then, as a string, it was deficient, because it didn't include the location from which the selected code was obtained.

So, I think there is an opportunity to redefine this codeSelection argument as a `string | UserContext` type, where `UserContext` can support all the different strategies that the user might employ to send context to Navie. 

This would include, but not necessarily be limited to:

- Code selections. A code selection is a range (start..end) from a file, plus the specific code from that range.
- Files. A file is just a file location. Navie should use the context API to retrieve the contents of that file. This operation is already supported in the context lookup facility. 

```
    type ContextV1.ContextRequest = {
        tokenCount: number;
        vectorTerms: string[]; <-- Leave this blank to fetch locations instead.
    };
    type ContextFilters = {
        recent?: boolean;
        locations?: string[]; <-- Can be populated to fetch the full contents of specified files.
        itemTypes?: ContextItemType[];
        labels?: ContextLabel[];
        exclude?: string[];
        include?: string[];
    };
    type ContextRequest = ContextV1.ContextRequest & ContextFilters;
```

Reply:
> Lookup by location is already supported like this:

Yes, I saw this, and a ContextRequest is ultimately the way file contents get retrieved. As far as I can tell, though, the existing interface for an Explain request doesn't allow the client to pass a ContextRequest. Instead, the pieces of the ContextRequest are passed as function parameters, and then get assembled into filters here: https://github.com/getappmap/appmap-js/blob/main/packages/navie/src/services/context-service.ts#L39 .

Reply:
You can check out the generate agent to see how it fetches full file contents

Reply:
I guess I'm confused by this comment. The Explain agent uses the `searchContext` method of `ContextService`, which has this signature:
```
  async searchContext(
    options: AgentOptions,
    tokensAvailable: () => number,
    additionalTerms: string[] = [],
    locations?: string[]
  ): Promise<void> 
```
This sets the passed `locations` on a `ContextFilter`, which it then passes to `this.lookupContextService.lookupContext`.

The Generate agent uses `ContextService.locationContext`, which has this signature:
```
  async locationContext(fileNames: string[]): Promise<void>
```
`locationContext` sets `fileNames` as the `locations` on a `ContextFilter`, then it too calls `this.lookupContextService.lookupContext`.

Aren't these doing the same thing to fetch the full file contents? I understood your suggestion about creating `UserContextItem` to encapsulate all the different kinds of context a user can provide, and then passing an array of those, rather than having `codeSelection` and `locations` be separate arguments. Can't this array then be passed to `searchContext` as part of `AgentOptions`?